### PR TITLE
Add benchmarks for UTF16 decoding

### PR DESCRIPTION
**Pull request by valeriyvan**
Adds benchmarks for UTF16 decoding

Partially resolves [SR-8905](https://bugs.swift.org/browse/SR-8905)

### CODE REVIEW

File: benchmark/single-source/UTF16Decode.swift [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  }
+}
+
+@inline(never)
+public func run_UTF16Decode_InitFromData(_ N: Int) {
+  let input = allStringsData
+  for _ in 0..<200*N {
+    blackHole(String(data: input, encoding: .utf16))
+  }
+}
```
Comment (xwu) [inline benchmark/single-source/UTF16Decode.swift:108]:
```suggestion
}

```

File: benchmark/single-source/UTF16Decode.swift [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+@inline(never)
+public func run_UTF16Decode_InitFromData_ascii(_ N: Int) {
+  let input = asciiData
+  for _ in 0..<1_000*N {
+    blackHole(String(data: input, encoding: .utf16))
+  }
+}
```
Comment (xwu) [inline benchmark/single-source/UTF16Decode.swift:123]:
```suggestion
}

```

File: benchmark/single-source/UTF16Decode.swift [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+let allStringsCustomNoncontiguous = CustomNoncontiguousCollection(allStringsCodeUnits)
+let asciiCustomNoncontiguous = CustomNoncontiguousCollection(Array(ascii.utf16))
+
+@inline(never)
+public func run_UTF16Decode_InitFromCustom_contiguous(_ N: Int) {
+  let input = allStringsCustomContiguous
+  for _ in 0..<200*N {
+    blackHole(String(decoding: input, as: UTF16.self))
+  }
+}
```
Comment (xwu) [inline benchmark/single-source/UTF16Decode.swift:187]:
```suggestion
}

```

File: benchmark/single-source/UTF16Decode.swift [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+
+@inline(never)
+public func run_UTF16Decode_InitFromCustom_noncontiguous(_ N: Int) {
+  let input = allStringsCustomNoncontiguous
+  for _ in 0..<200*N {
+    blackHole(String(decoding: input, as: UTF16.self))
+  }
+}
```
Comment (xwu) [inline benchmark/single-source/UTF16Decode.swift:202]:
```suggestion
}

```

**Review by xwu — COMMENTED** [review]
@swift-ci Please benchmark

**Comment by xwu** [comment]
@swift-ci Please benchmark

**Comment by xwu** [comment]
@swift-ci Please benchmark

**Review by xwu — COMMENTED** [review]
The suggestions on how to modify the benchmarks are pretty helpful.

**Comment by valeriyvan** [comment]
> The suggestions on how to modify the benchmarks are pretty helpful.

Suggestions absolutely make sense. But. Benchmarks which were added are copycats of UTF8 benchmarks. They use same names. They use same test strings. I think, it's very interesting to see that `UTF8Decode_InitDecoding` takes 195μs and `UTF16Decode_InitDecoding` takes 84358μs which is several orders of magnitude slower. If you agree comparison of these benchmarks is important for future, tests in this PR should follow names from UTF8Decode benchmark. For the same reason benchmarks shouldn't be optimised to be executed less times. Otherwise we lose base of comparison to UTF8. I am going to add UTF32Decode as well, as separate benchmark. So we will have overview of all speed of all used UTF conversions.

What do you think of it, @xwu?

**Comment by xwu** [comment]
No, those are legacy names. Benchmarks now have a different naming scheme, and all new benchmarks should adhere to them. The workload and setup warnings should be adhered to as well precisely because it is important to have reliable benchmarks that can be compared over time.

**Comment by eeckstein** [comment]
Is it possible to reduce the number of new benchmarks to a set which contains a representative coverage of the algorithms/code which it should test?
I know, it's just a copy of the UTF8 benchmark file (but also for this we should think of reducing the number of benchmarks).

If we end up adding benchmarks for all combinations of all language features, the benchmark runtime just gets out of bounds.

Note: it's possible to add the ".skip" benchmark tag, which excludes a benchmark from the regular run, but still enables someone to run it locally.

**Comment by valeriyvan** [comment]
> Is it possible to reduce the number of new benchmarks to a set which contains a representative coverage of the algorithms/code which it should test? I know, it's just a copy of the UTF8 benchmark file (but also for this we should think of reducing the number of benchmarks).
> 
> If we end up adding benchmarks for all combinations of all language features, the benchmark runtime just gets out of bounds.
> 
> Note: it's possible to add the ".skip" benchmark tag, which excludes a benchmark from the regular run, but still enables someone to run it locally.

done

**Comment by valeriyvan** [comment]
ping

**Comment by eeckstein** [comment]
@swift-ci Please benchmark

**Comment by eeckstein** [comment]
> ⛔️⏱ | `UTF16Decode.initFromData` has setup overhead of 9 μs (5.6%). <br><sub> Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`

It would be good to force initialization of the global variables in the setUpFunction, e.g. add `blackHole(allStringsData)` to the setup function(s)

**Comment by valeriyvan** [comment]
> > ⛔️⏱ | `UTF16Decode.initFromData` has setup overhead of 9 μs (5.6%).  Move initialization of benchmark data to the `setUpFunction` registered in `BenchmarkInfo`
> 
> It would be good to force initialization of the global variables in the setUpFunction, e.g. add `blackHole(allStringsData)` to the setup function(s)

done

**Comment by eeckstein** [comment]
@swift-ci benchmark

**Comment by eeckstein** [comment]
@swift-ci smoke test

**Comment by eeckstein** [comment]
@swift-ci smoke test linux

**Comment by valeriyvan** [comment]
> @swift-ci smoke test linux

Should I bother about failed linux test?

**Comment by AnthonyLatsis** [comment]
@swift-ci please smoke test Linux

**Comment by eeckstein** [comment]
@valeriyvan The linux failure was unrelated. Thanks for the contribution!

**Pull request merged**